### PR TITLE
feat/dpo lora

### PR DESCRIPTION
**Pull request by danielsanchezaran**
# feat: LoRA adapter support for DPO fine-tuning

## Summary

This PR adds LoRA (Low-Rank Adaptation) support to the DPO preference-optimization pipeline,
enabling parameter-efficient fine-tuning of the DiT decoder without modifying the full model
weights. It also includes DPO training quality improvements (variance reduction, drift tracking)
and fixes for ONNX export with LoRA weights.

---

### DiT decoder — unfused attention for LoRA compatibility

`diffusion_planner/model/module/dit.py`

`nn.MultiheadAttention` fuses Q/K/V projections into a single `in_proj_weight` tensor.
PEFT cannot target individual projections with this layout. The self-attention and
cross-attention layers in `DiTBlock` are replaced with `UnfusedMultiheadAttention`, which
exposes separate `q_proj` / `k_proj` / `v_proj` / `out_proj` `Linear` modules that match
the LoRA adapter target pattern.

`model_utils.load_model` applies `_migrate_fused_attention()` to handle checkpoints saved
before this refactor (`in_proj_weight` layout), so existing trained models load without
requiring retraining.

`requirements.txt`: adds `peft` as a dependency.

### LoRA utilities (`preference_optimization/lora_utils.py`)

- `apply_lora(model, rank, alpha, dropout, target_modules)` — wraps the model with a PEFT
  `LoraConfig` targeting the unfused attention projections and the DiT `fc` layers.
- `load_lora_checkpoint(model, lora_dir, is_trainable)` — loads a saved adapter via
  `PeftModel.from_pretrained`; `is_trainable=True` keeps gradients enabled for continued
  training.
- `merge_lora_and_unload(model)` — merges `W_merged = W_base + (alpha/r) * B @ A` and
  returns a plain `nn.Module` ready for ONNX export.

### DPO training — resume and drift tracking (`preference_optimization/train_dpo.py`, `trainer.py`)

**Resume from checkpoint:**
`setup_experiment()` detects an existing LoRA adapter directory (`lora_latest/` symlink or
`lora_epoch_NNN/`) alongside `--model_path` and copies it to `run_dir/lora_seed/`. On
`main()` startup, if `lora_seed/` is present, training resumes via
`load_lora_checkpoint(..., is_trainable=True)` instead of initialising fresh zero-delta
weights. The `adapter_config.json` baked into the saved adapter supplies rank/alpha/dropout,
so `--lora_rank` / `--lora_alpha` / `--lora_dropout` are ignored on resume.

**Epoch-1 drift metric:**
Before the first training epoch begins, the trainer saves deterministic (zero-noise)
trajectory baselines for each annotated sample to `run_dir/epoch1_baselines.npz`. After
every subsequent epoch the mean/std/max ADE between the current model output and these
baselines is computed and displayed in both stdout and the annotation GUI sidebar under
"Model drift vs epoch 1". This makes it easy to verify that DPO training is actually
changing model behaviour.

### DPO loss — variance reduction (`preference_optimization/dpo_loss.py`)

- The same `(noise, t)` sample is now shared between the winner and loser forward passes in
  each step of the K-step Monte Carlo estimate. Correlated noise cancels in the difference,
  substantially reducing loss variance.
- K is increased to 8 for a lower-variance estimator.
- A fresh `disable_adapter()` context manager is created on each inner iteration to avoid
  `AttributeError` on context-manager reuse.

### ONNX export — LoRA weight merging (`ros_scripts/torch2onnx.py`)

**LoRA detection fix:**
The previous implementation only checked for `adapter_config.json` next to the `.pth` file
(single-directory layout). The DPO trainer saves the base checkpoint as `run_dir/latest.pth`
and adapters as `run_dir/lora_epoch_NNN/`, so the adapter was silently not found and ONNX
export used untrained base weights. The search now also probes `run_dir/lora_latest/` (symlink
maintained by the trainer) and the highest-numbered `lora_epoch_NNN/` directory.

**Simplified ONNX validation:**
When `--use-simplify` is passed, the same ORT vs PyTorch numerical comparison that runs on
the base `.onnx` is now also run on the `_simplified.onnx` produced by `onnxsim`. If
simplification fails for any reason, `simplified_path` is set to `None` and the validation
block is skipped gracefully.

---

## Test plan

- [ ] Train a short DPO run with `--use_lora` from scratch and confirm `lora_epoch_NNN/`
      directories and `lora_latest` symlink are created.
- [ ] Resume the run from the saved adapter and confirm loss continues from where it left off
      (no weight reset).
- [ ] Export with `python3 ros_scripts/torch2onnx.py <run_dir> --use-simplify` and confirm:
      - "LoRA weights merged from …" is printed (not "No LoRA adapter found").
      - Both base and simplified ONNX pass the ORT numerical comparison (`Close? True`).
- [ ] Load an old checkpoint (pre-unfused-attention) and confirm it loads without errors via
      `_migrate_fused_attention`.
- [ ] Run the DPO unit tests: `python3 preference_optimization/test_fde_standalone.py`.

Related ticket w/ videos: https://tier4.atlassian.net/browse/T4DEV-50357

### CODE REVIEW

File: diffusion_planner/diffusion_planner/model/module/dit.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 import torch.nn as nn
 from timm.models.layers import Mlp
 
+from diffusion_planner.model.module.attention import UnfusedMultiheadAttention
```
Comment (SakodaShintaro) [inline diffusion_planner/diffusion_planner/model/module/dit.py]: Where can I find this?

**Review by SakodaShintaro — CHANGES_REQUESTED** [review]
Hmm, changing `diffusion_planner/model/module/dit.py` is not ideal for me.

In my understanding, `peft` can be applied to a model without changing the original code.

https://github.com/huggingface/peft?tab=readme-ov-file#quickstart

Could you rethink the way to use `peft`?

File: diffusion_planner/diffusion_planner/model/module/dit.py (hunk 1 above)
Comment (danielsanchezaran) [reply diffusion_planner/diffusion_planner/model/module/dit.py]: This was a file that got deleted, I have removed the import in [8297e1f](https://github.com/tier4/Diffusion-Planner/pull/26/commits/8297e1f0475d14b891a886646dfcea6f1f2f1a18)

**Comment by danielsanchezaran** [comment]
> Hmm, changing `diffusion_planner/model/module/dit.py` is not ideal for me.
> 
> In my understanding, `peft` can be applied to a model without changing the original code.
> 
> https://github.com/huggingface/peft?tab=readme-ov-file#quickstart
> 
> Could you rethink the way to use `peft`?

Thank you for your reply, I have removed changes from dit.py and implemented it a different way now in [8297e1f](https://github.com/tier4/Diffusion-Planner/pull/26/commits/8297e1f0475d14b891a886646dfcea6f1f2f1a18)

**Comment by danielsanchezaran** [comment]
@SakodaShintaro changes to DiT and torch2onnx were removed. Instead I added a merge script that lets us merge model .pth files with lora weights into a single .pth

**Review by SakodaShintaro — APPROVED** [review]
